client/src/components/chat/chatTab.js

```diff
@@ -51,7 +54,37 @@ export default class chatTab extends Component {
             if (data.chatId !== this.props.chatId){
                 return;
             }
-            self.setState( { messages : self.state.messages.concat([data]) } );
+
+            this.setState( {
+                showIsTyping: '',
+                messages : self.state.messages.concat([data])
+            })
+        })
+
+        this.socket.on(`server:userConnected`, data => {
+
+            if (!this.ignoredOwnConnectMessage){
+                this.ignoredOwnConnectMessage = true
+                return
+            }
+
+            let newUserMsg = {
+                clientName: "system",
+                text: `${data.clientName} ${Strings.chat.hasJoined}`,
+                createdAd: Moment(),
+                isMe: false
+            }
+
+            this.setState({messages: this.state.messages.concat([ newUserMsg ])})
+        })
+
+        this.socket.on('server:clientIsTyping', data => {
+            if (data.clientId !== this.state.clientId){
+                this.setState({showIsTyping:`${data.clientName} ${Strings.chat.isTyping}`});
+                setTimeout( () => {
```

Comment:
hmm maybe some like save the timeout reference (setTimeout's return value) and do a clearTimeout() on it when you get a new event?